diff -up --recursive mod_nss-1.0.11/Makefile.am mod_nss-1.0.11.test/Makefile.am
--- mod_nss-1.0.11/Makefile.am	2015-09-21 16:41:05.789937357 -0400
+++ mod_nss-1.0.11.test/Makefile.am	2015-09-21 16:42:38.112473063 -0400
@@ -103,12 +103,6 @@ check:
 	cd test;	\
 	rm -rf work;	\
 	nosetests -v test_cipher.py;	\
-	./setup.sh;	\
-	nosetests -v test.py;	\
-	sleep 5;	\
-	rm -rf work;	\
-	./setup.sh sql:;	\
-	DBPREFIX=sql: nosetests -v test.py;	\
 	cd ..
 
 .PHONY: all test clean
diff -up --recursive mod_nss-1.0.11/test/test_request.py mod_nss-1.0.11.test/test/test_request.py
--- mod_nss-1.0.11/test/test_request.py	2014-12-02 13:59:03.000000000 -0500
+++ mod_nss-1.0.11.test/test/test_request.py	2015-09-21 16:42:25.563672095 -0400
@@ -4,8 +4,8 @@
 import socket
 import requests
 import urlparse
-from urllib3.util import get_host
-from urllib3.connectionpool import HTTPConnectionPool, HTTPSConnectionPool
+from requests.packages.urllib3.util import get_host
+from requests.packages.urllib3.connectionpool import HTTPConnectionPool, HTTPSConnectionPool
 import logging
 
 # Don't bend over backwards for ssl support, assume it is there.
@@ -107,6 +107,9 @@ class MyVerifiedHTTPSConnection(HTTPSCon
     ca_certs = None
     client_cipher = None
 
+    # A gross oversimplification but we aren't testing this
+    is_verified = True
+
     def set_cert(self, key_file=None, cert_file=None,
                  cert_reqs='CERT_NONE', ca_certs=None):
         ssl_req_scheme = {
