diff -rupN mod_nss-1.0.10.orig/nss_engine_init.c mod_nss-1.0.10/nss_engine_init.c
--- mod_nss-1.0.10.orig/nss_engine_init.c	2016-01-15 14:53:24.000000000 -0800
+++ mod_nss-1.0.10/nss_engine_init.c	2016-01-18 14:52:41.000000000 -0800
@@ -1236,9 +1236,8 @@ static void nss_init_server_certs(server
 #endif
         {
             /*
-             * Since this is a 'fatal' error, regardless of whether this
-             * particular invocation is from a 'server' object or a 'proxy'
-             * object, issue all error message(s) as appropriate.
+             * This is a 'fatal' error since a 'nickname' is required for
+             * a 'server' object; log error message(s) as appropriate.
              */
             if ((mctx->sc->enabled == TRUE) &&
                 (mctx->sc->server) &&
@@ -1247,13 +1246,6 @@ static void nss_init_server_certs(server
                     "NSSEngine on; no certificate nickname provided by NSSNickname.");
             }
 
-            if ((mctx->sc->proxy_enabled == TRUE) &&
-                (mctx->sc->proxy) &&
-                (mctx->sc->proxy->nickname == NULL)) {
-                ap_log_error(APLOG_MARK, APLOG_ERR, 0, s,
-                    "NSSProxyEngine on; no certificate nickname provided by NSSProxyNickname.");
-            }
-
             nss_die();
         }
 
