diff -rupN mod_nss-1.0.8.orig/nss_engine_vars.c mod_nss-1.0.8/nss_engine_vars.c
--- mod_nss-1.0.8.orig/nss_engine_vars.c	2014-06-11 10:38:12.000000000 -0700
+++ mod_nss-1.0.8/nss_engine_vars.c	2014-06-12 17:42:27.000000000 -0700
@@ -575,7 +575,8 @@ static char *nss_var_lookup_nss_cert_PEM
     len = strlen(tmp);
     while (tmp[i] != '\0') {
         if (tmp[i] == '\r') {
-            memmove(&tmp[i], &tmp[i+1], 1+(len - i));
+            memmove(&tmp[i], &tmp[i+1], (len - i));
+            len--;
         }
         i++;
     }
