From 770139456144c576335f407f990f14129e70d4c2 Mon Sep 17 00:00:00 2001
From: Rob Crittenden <rcritten@redhat.com>
Date: Thu, 20 Feb 2014 17:22:32 -0500
Subject: [PATCH] Modify gencert to create a password-less database

---
 gencert.in | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/gencert.in b/gencert.in
index f51ee3e..dd3e0c3 100755
--- a/gencert.in
+++ b/gencert.in
@@ -82,12 +82,11 @@ fi
 
 DEST=$1
 
-echo "httptest" > $DEST/pw.txt
+echo -e "\n" > $DEST/pw.txt
 
 echo ""
 echo "#####################################################################"
-echo "Generating new server certificate and key database. The password"
-echo "is httptest"
+echo "Generating new server certificate and key database."
 echo "#####################################################################"
 $CERTUTIL -N -d $DEST -f $DEST/pw.txt
 
@@ -183,8 +182,4 @@ echo "#####################################################################"
 rm $DEST/pw.txt
 rm $DEST/noise
 
-echo ""
-echo "The database password is httptest"
-echo ""
-
 exit 0
-- 
1.7.11.7

